app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/TagsSelectorPresenter.kt

```diff
@@ -0,0 +1,145 @@
+package space.taran.arknavigator.mvp.presenter.adapter
+
+import android.util.Log
+import space.taran.arknavigator.mvp.model.dao.ResourceId
+import space.taran.arknavigator.mvp.model.repo.TagsStorage
+import space.taran.arknavigator.mvp.view.ResourcesView
+import space.taran.arknavigator.utils.Popularity
+import space.taran.arknavigator.utils.TAGS_SELECTOR
+import space.taran.arknavigator.utils.Tag
+import java.lang.AssertionError
+
+class TagsSelectorPresenter(
+    val viewState: ResourcesView,
+    val onSelectionChangeListener: (Set<ResourceId>) -> Unit
+) {
+    private var resources: List<ResourceId>? = null
+    private var storage: TagsStorage? = null
+
+    private var included = mutableSetOf<Tag>()
+    private var excluded = mutableSetOf<Tag>()
+
+    var selection = setOf<ResourceId>()
+
+    //this data is used by TagsSelectorAdapter
+    var includedTags = listOf<Tag>()
+    var excludedTags = listOf<Tag>()
+    var availableTags = listOf<Tag>()
+    var unavailableTags = listOf<Tag>()
+    var isClearBtnVisible = false
+
+    fun init(resources: List<ResourceId>, storage: TagsStorage) {
+        this.resources = resources
+        this.storage = storage
+    }
+
+    fun onTagClick(tag: Tag) {
+        if (excluded.contains(tag) || included.contains(tag)) {
+            uncheckTag(tag)
+        } else {
+            includeTag(tag)
+        }
+    }
+
+    fun onTagLongClick(tag: Tag) {
+        if (!excluded.contains(tag) || included.contains(tag)) {
+            excludeTag(tag)
+        } else {
+            uncheckTag(tag)
+        }
+    }
+
+    fun onClearClick() {
+        (included + excluded).forEach { uncheckTag(it, needToCalculate = false) }
+        calculateTagsAndSelection()
+    }
+
+    fun calculateTagsAndSelection() {
+        if (storage == null || resources == null)
+            return
+
+        val allTags = storage!!.listAllTags()
```

Comment:
This is not equivalent to `resources.flatMap { storage!!.getTags(it) }` (see other comment about `listAllTags`). Maybe we could design `TagsStorage` as more granular, so we would create one tags storage per every favorite, but I am not sure that we would benefit from it.